tensordict/base.py

```diff
@@ -2374,6 +2452,110 @@ def saved_path(self):
             f"The tensordict has no saved path (memmap={self.is_memmap()}, path={self._memmap_prefix})."
         )
 
+    def consolidate(
+        self,
+        *,
+        num_threads=0,
+        filename: Path | str | None = None,
+        device: torch.device | None = None,
+    ) -> None:
+        """Consolidates the tensordict content in a single storage for fast serialization.
+
+        Keyword Args:
+            num_threads (integer, optional): the number of threads to use for populating
+                the storage.
+            filename (Path, optional): an optional file path for a memory-mapped tensor
+                to use as a storage for the tensordict.
+            device (torch.device, optional): an optional device where the storage must be
+                instantiated.
+
+        Examples:
+            >>> import pickle
+            >>> import tempfile
+            >>> import torch
+            >>> import tqdm
+            >>> from torch.utils.benchmark import Timer
+            >>> from tensordict import TensorDict
+            >>> data = TensorDict({"a": torch.zeros(()), "b": {"c": torch.zeros(())}})
+            >>> data_consolidated = data.consolidate()
+            >>> # check that the data has a single data_ptr()
+            >>> assert torch.tensor([
+            ...     v.untyped_storage().data_ptr() for v in data_c.values(True, True)
+            ... ]).unique().numel() == 1
+            >>> # Serializing the tensordict will be faster with data_consolidated
+            >>> with open("data.pickle", "wb") as f:
+            ...    print("regular", Timer("pickle.dump(data, f)", globals=globals()).adaptive_autorange())
+            >>> with open("data_c.pickle", "wb") as f:
+            ...     print("consolidated", Timer("pickle.dump(data_consolidated, f)", globals=globals()).adaptive_autorange())
+
+
+        """
+        flat_dict = dict(zip(*self._items_list(leaves_only=True, include_nested=True)))
+        flat_size = [v.element_size() * v.numel() for v in flat_dict.values()]
+        if filename is None:
+            storage = torch.empty(
+                sum(flat_size),
+                dtype=torch.bool,
+                device=device if device else self.device,
+            )
+        else:
+            if device not in (torch.device("cpu"), None):
+                raise RuntimeError(
+                    "device and filename are mutually exclusive arguments."
+                )
+            storage = torch.from_file(
+                str(filename), size=sum(flat_size), dtype=torch.bool, shared=True
+            )
+
+        flat_metadata = {}
+
+        offsets = torch.tensor([0] + flat_size).cumsum(0)
+
+        def assign(k, v, start, stop, storage=storage):
+            offset = v.storage_offset()
+            stride = v.stride()
+            if offset or stride != 1:
+                content = v.clone().untyped_storage()
```

Comment:
It would be nice to be able to get a view of the storage directly
Do we have that?
cc @albanD @mikaylagawarecki

Reply:
Well a "view of a storage" has a name, it's a "Tensor" :D 
If you need a contiguous chunk of memory containing these values, you will need a copy indeed and clone() is the way to go!

Note that you might want to pass in memory_format=torch.contiguous_format if you want a row major storage. Otherwise, you might get a channel last here as the default is to preserve the memory format.